{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df4f6ab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append(\"/mnt/lareaulab/reliscu/code\")\n",
    "\n",
    "from parse_gtf import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "cc2be6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "psi = pd.read_csv(\"data/tasic_2018_ALM_STAR_SyntheticDataset1_20pcntCells_25SD_200samples_SJ_pseudobulk_min_observed0.25_minPsi0.1_PSI.csv\", index_col=0)\n",
    "top_qval_mods_df = pd.read_csv(\"data/tasic_2018_ALM_STAR_donor_cell_type_pseudobulk_pairwise_DE_genes_dream_20pcntCells_25SD_200samples_log2_pseudobulk_PosFDR_top_Qval_modules.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c96847b3",
   "metadata": {},
   "source": [
    "### Add gene names to data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f11ee42a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = \"/mnt/lareaulab/reliscu/data/GENCODE/GRCm39/gencode.vM35.annotation.gtf\"\n",
    "gtf = gtf_parse(gtf_file)\n",
    "gtf_subset = gtf.loc[gtf['feature'].isin([\"gene\"])]\n",
    "attrs = gtf_subset[\"attribute\"].apply(extract_attributes)\n",
    "attrs_df = attrs.apply(pd.Series)\n",
    "gtf_parsed = pd.concat([gtf_subset.drop(columns=[\"attribute\"]), attrs_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "054bd5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_parsed['gene_id'] = gtf_parsed['gene_id'].str.split(\".\").str[0]\n",
    "psi['gene_id'] = psi.index.str.split(\"_\").str[0]\n",
    "psi['exon_id'] = psi.index.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "b4c7976d",
   "metadata": {},
   "outputs": [],
   "source": [
    "psi_anno = pd.merge(gtf_parsed[['gene_id', 'gene_name']], psi, on=\"gene_id\", how=\"right\")\n",
    "psi_anno = psi_anno.set_index(\"exon_id\").rename_axis(None)\n",
    "psi_anno = psi_anno.drop(columns=[\"gene_id\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80e4980c",
   "metadata": {},
   "source": [
    "### Calc. corr between ME and exon PSI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "097d4e1b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8c9b19c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each cell type module, correlate its eigengene with exon PSI:\n",
    "\n",
    "eigengene_vs_psi <- lapply(seq_along(filtered_mes$Cell_type), function(i) {\n",
    "    me <- fread(filtered_mes$ME_path[i], data.table=FALSE)\n",
    "    me <- me[,grep(paste0(\"^\", filtered_mes$Module[i], \"$\"), colnames(me))]\n",
    "    me_psi_cor <- apply(psi[,-1], 1, function(exon) {\n",
    "        cor_test <- cor.test(exon, me, alternative=\"greater\")\n",
    "        data.frame(Cor=cor_test$estimate, Pval=cor_test$p.value)\n",
    "    })\n",
    "    data.frame(Cell_type=filtered_mes$Cell_type[i], \n",
    "               Gene_name=gene_names, \n",
    "               Gene_ID=gene_ids, \n",
    "               Exon=exon_ids, \n",
    "               Exon_Mean_Expr=mean_expr,\n",
    "               do.call(rbind, me_psi_cor), \n",
    "               row.names=NULL)\n",
    "})\n",
    "names(eigengene_vs_psi) <- filtered_mes$Cell_type"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
