{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "df4f6ab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "\n",
    "sys.path.append(\"/mnt/lareaulab/reliscu/code\")\n",
    "\n",
    "from parse_gtf import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b8fde77",
   "metadata": {},
   "outputs": [],
   "source": [
    "psi_data = \"SyntheticDataset1_20pcntCells_35SD_200samples_SJ_pseudobulk_min_observed0.25_minPsi0.1_PSI\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc2be6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "psi = pd.read_csv(f\"data/tasic_2018_ALM_STAR_{psi_data}.csv\", index_col=0)\n",
    "top_qval_mods_df = pd.read_csv(\"data/tasic_2018_ALM_STAR_donor_cell_type_pseudobulk_pairwise_DE_genes_dream_20pcntCells_35SD_200samples_log2_pseudobulk_PosBC_top_Qval_modules.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c96847b3",
   "metadata": {},
   "source": [
    "### Add gene names to PSI data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f11ee42a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse GTF attribute column\n",
    "gtf_file = \"/mnt/lareaulab/reliscu/data/GENCODE/GRCm39/gencode.vM35.annotation.gtf\"\n",
    "gtf = gtf_parse(gtf_file)\n",
    "gtf_subset = gtf.loc[gtf['feature'].isin([\"gene\"])]\n",
    "attrs = gtf_subset[\"attribute\"].apply(extract_attributes)\n",
    "attrs_df = attrs.apply(pd.Series)\n",
    "gtf_parsed = pd.concat([gtf_subset.drop(columns=[\"attribute\"]), attrs_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "054bd5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get PSI and GTF data ready to merge on gene IDs\n",
    "gtf_parsed['gene_id'] = gtf_parsed['gene_id'].str.split(\".\").str[0]\n",
    "psi['gene_id'] = psi.index.str.split(\"_\").str[0]\n",
    "psi['exon_id'] = psi.index.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b4c7976d",
   "metadata": {},
   "outputs": [],
   "source": [
    "psi_anno = pd.merge(gtf_parsed[['gene_id', 'gene_name']], psi, on=\"gene_id\", how=\"right\")\n",
    "psi_anno = psi_anno.set_index(\"exon_id\").rename_axis(None)\n",
    "psi_anno = psi_anno.drop(columns=[\"gene_id\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80e4980c",
   "metadata": {},
   "source": [
    "### Calc. corr between ME and exon PSI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87933e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_df = pd.DataFrame(\n",
    "    columns=[\"Gene\"] + top_qval_mods_df['Cell_type'].tolist(), \n",
    "    index=psi_anno.index\n",
    ")\n",
    "corr_df['Gene'] = psi_anno['gene_name'] \n",
    "\n",
    "for i, row in top_qval_mods_df.iterrows():\n",
    "    ctype = row['Cell_type']\n",
    "\n",
    "    mod_df = pd.read_csv(row['ME_path'])\n",
    "    mod_eig = mod_df.set_index(\"Sample\")[row['Module']]\n",
    "    mod_eig = pd.to_numeric(mod_eig, errors=\"coerce\")\n",
    "    \n",
    "    corrs = psi_anno.iloc[:, 1:].corrwith(mod_eig, axis=1)\n",
    "    corr_df[ctype] = corrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3805f50d",
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_df.to_csv(f\"data/tasic_2018_ALM_STAR_{psi_data}_exon_corr.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
